File: ios/Packages/RideLinkCore/Sources/RideLinkCore/Protocol/PlaybackCodec.swift
import Foundation

/// PROTOCOL §3 Playback group (Phase 5).
public enum PlaybackMessageTypes {
    public static let play = "PLAY"
    public static let pause = "PAUSE"
    public static let resume = "RESUME"
    public static let seek = "SEEK"
    public static let next = "NEXT"
    public static let previous = "PREVIOUS"
    public static let positionReport = "POSITION_REPORT"
    public static let playbackState = "PLAYBACK_STATE"

    public static let all: Set<String> = [play, pause, resume, seek, next, previous, positionReport, playbackState]

    /// The subset that carries a `PlaybackCommandHeader` and is therefore ordered by `command_seq`.
    public static let commands: Set<String> = [play, pause, resume, seek, next, previous]
}

/// Why a playback payload was refused. Recorded in diagnostics; never sent to the peer verbatim.
///
/// Raw values match the Kotlin enum constant names exactly, since
/// `protocol/vectors/playback-messages/` names them that way and both platforms read the same file.
public enum PlaybackMessageRejection: String, Sendable, Equatable {
    case unknownType = "UNKNOWN_TYPE"
    case missingField = "MISSING_FIELD"
    case wrongFieldType = "WRONG_FIELD_TYPE"
    case malformedContentHash = "MALFORMED_CONTENT_HASH"
    case malformedQueueItemId = "MALFORMED_QUEUE_ITEM_ID"
    case malformedPeerId = "MALFORMED_PEER_ID"
    case commandSeqOutOfRange = "COMMAND_SEQ_OUT_OF_RANGE"
    case revisionOutOfRange = "REVISION_OUT_OF_RANGE"
    case sessionTimeOutOfRange = "SESSION_TIME_OUT_OF_RANGE"
    case positionOutOfRange = "POSITION_OUT_OF_RANGE"
    case rateOutOfRange = "RATE_OUT_OF_RANGE"
}

/// Parses, bounds-checks and encodes PROTOCOL §5's playback messages. Total and non-trapping,
/// mirroring `TransferCodec`/`AudioStateCodec`/`VoiceSignalCodec` — a malformed frame is dropped and
/// the control connection survives, because the framing was intact and only this message's shape was
/// wrong.
///
/// Mirrors Android `core.protocol.PlaybackCodec`; both run `protocol/vectors/playback-messages/`.
public enum PlaybackCodec {
    public enum Result: Sendable, Equatable {
        case parsed(PlaybackMessage)
        case rejected(PlaybackMessageRejection)
    }

    public static let fieldCommandSeq = "command_seq"
    public static let fieldEffectiveAtSessionUs = "effective_at_session_us"
    public static let fieldIssuedBy = "issued_by"
    public static let fieldQueueRevision = "queue_revision"
    public static let fieldTrackHash = "track_hash"
    public static let fieldPositionMs = "position_ms"
    public static let fieldTargetPositionMs = "target_position_ms"
    public static let fieldQueueItemId = "queue_item_id"
    public static let fieldAtSessionUs = "at_session_us"
    public static let fieldPlaying = "playing"
    public static let fieldPlaybackRate = "playback_rate"

    public static func parse(type: String, payload: [String: JSONValue]) -> Result {
        switch type {
        case PlaybackMessageTypes.play: return parsePlay(payload)
        case PlaybackMessageTypes.pause: return parsePauseOrResume(payload, resume: false)
        case PlaybackMessageTypes.resume: return parsePauseOrResume(payload, resume: true)
        case PlaybackMessageTypes.seek: return parseSeek(payload)
        case PlaybackMessageTypes.next: return parseStep(payload, next: true)
        case PlaybackMessageTypes.previous: return parseStep(payload, next: false)
        case PlaybackMessageTypes.positionReport: return parsePositionReport(payload)
        case PlaybackMessageTypes.playbackState: return parsePlaybackState(payload)
        default: return .rejected(.unknownType)
        }
    }

    public static func wireType(_ message: PlaybackMessage) -> String {
        switch message {
        case .play: return PlaybackMessageTypes.play
        case .pause: return PlaybackMessageTypes.pause
        case .resume: return PlaybackMessageTypes.resume
        case .seek: return PlaybackMessageTypes.seek
        case .next: return PlaybackMessageTypes.next
        case .previous: return PlaybackMessageTypes.previous
        case .positionReport: return PlaybackMessageTypes.positionReport
        case .playbackState: return PlaybackMessageTypes.playbackState
        }
    }

    /// The outbound side of `parse` — the shape lives here, once, shared by both directions.
    public static func encode(_ message: PlaybackMessage) -> [String: JSONValue] {
        switch message {
        case .play(let header, let trackHash, let positionMs, let queueItemId):
            return encodeHeader(header)
                .merging([
                    fieldTrackHash: .string(trackHash.value),
                    fieldPositionMs: .number(Double(positionMs)),
                    fieldQueueItemId: .string(queueItemId),
                ]) { _, new in new }
        case .pause(let header, let positionMs), .resume(let header, let positionMs):
            return encodeHeader(header).merging([fieldPositionMs: .number(Double(positionMs))]) { _, new in new }
        case .seek(let header, let targetPositionMs):
            return encodeHeader(header).merging([fieldTargetPositionMs: .number(Double(targetPositionMs))]) { _, new in new }
        case .next(let header), .previous(let header):
            return encodeHeader(header)
        case .positionReport(let trackHash, let positionMs, let atSessionUs, let playing, let playbackRate):
            return [
                fieldTrackHash: .string(trackHash.value),
                fieldPositionMs: .number(Double(positionMs)),
                fieldAtSessionUs: .number(Double(atSessionUs)),
                fieldPlaying: .bool(playing),
                fieldPlaybackRate: .number(playbackRate),
            ]
        case .playbackState(
            let commandSeq, let queueRevision, let trackHash, let queueItemId, let positionMs, let playing, let atSessionUs
        ):
            return [
                fieldCommandSeq: .number(Double(commandSeq)),
                fieldQueueRevision: .number(Double(queueRevision)),
                fieldTrackHash: trackHash.map { JSONValue.string($0.value) } ?? .null,
                fieldQueueItemId: queueItemId.map { JSONValue.string($0) } ?? .null,
                fieldPositionMs: .number(Double(positionMs)),
                fieldPlaying: .bool(playing),
                fieldAtSessionUs: .number(Double(atSessionUs)),
            ]
        }
    }

    private static func encodeHeader(_ header: PlaybackCommandHeader) -> [String: JSONValue] {
        [
            fieldCommandSeq: .number(Double(header.commandSeq)),
            fieldEffectiveAtSessionUs: .number(Double(header.effectiveAtSessionUs)),
            fieldIssuedBy: .string(header.issuedBy.value),
            fieldQueueRevision: .number(Double(header.queueRevision)),
        ]
    }

    // MARK: - Headers

    private enum HeaderResult {
        case ok(PlaybackCommandHeader)
        case bad(Result)
    }

    private static func parseHeader(_ payload: [String: JSONValue]) -> HeaderResult {
        guard let commandSeq = phase5LongField(payload, fieldCommandSeq) else {
            return .bad(missingOrWrongType(payload, fieldCommandSeq))
        }
        if commandSeq < 0 || commandSeq > PlaybackBounds.maxWireInt { return .bad(.rejected(.commandSeqOutOfRange)) }
        guard let effectiveAt = phase5LongField(payload, fieldEffectiveAtSessionUs) else {
            return .bad(missingOrWrongType(payload, fieldEffectiveAtSessionUs))
        }
        if effectiveAt < 0 || effectiveAt > PlaybackBounds.maxWireInt { return .bad(.rejected(.sessionTimeOutOfRange)) }
        guard let issuedByRaw = phase5StringField(payload, fieldIssuedBy) else {
            return .bad(missingOrWrongType(payload, fieldIssuedBy))
        }
        guard let issuedBy = phase5ParsePeerId(issuedByRaw) else { return .bad(.rejected(.malformedPeerId)) }
        guard let queueRevision = phase5LongField(payload, fieldQueueRevision) else {
            return .bad(missingOrWrongType(payload, fieldQueueRevision))
        }
        if queueRevision < 0 || queueRevision > PlaybackBounds.maxWireInt { return .bad(.rejected(.revisionOutOfRange)) }
        return .ok(
            PlaybackCommandHeader(
                commandSeq: commandSeq, effectiveAtSessionUs: effectiveAt, issuedBy: issuedBy, queueRevision: queueRevision
            )
        )
    }

    // MARK: - Per-type parsers

    private static func parsePlay(_ payload: [String: JSONValue]) -> Result {
        let header: PlaybackCommandHeader
        switch parseHeader(payload) {
        case .bad(let rejected): return rejected
        case .ok(let parsed): header = parsed
        }
        guard let hashRaw = phase5StringField(payload, fieldTrackHash) else { return missingOrWrongType(payload, fieldTrackHash) }
        guard let trackHash = ContentHash.parse(hashRaw) else { return .rejected(.malformedContentHash) }
        guard let positionMs = phase5LongField(payload, fieldPositionMs) else { return missingOrWrongType(payload, fieldPositionMs) }
        guard isValidPosition(positionMs) else { return .rejected(.positionOutOfRange) }
        guard let queueItemId = phase5StringField(payload, fieldQueueItemId) else { return missingOrWrongType(payload, fieldQueueItemId) }
        guard phase5IsUlid(queueItemId) else { return .rejected(.malformedQueueItemId) }
        return .parsed(.play(header: header, trackHash: trackHash, positionMs: positionMs, queueItemId: queueItemId))
    }

    private static func parsePauseOrResume(_ payload: [String: JSONValue], resume: Bool) -> Result {
        let header: PlaybackCommandHeader
        switch parseHeader(payload) {
        case .bad(let rejected): return rejected
        case .ok(let parsed): header = parsed
        }
        guard let positionMs = phase5LongField(payload, fieldPositionMs) else { return missingOrWrongType(payload, fieldPositionMs) }
        guard isValidPosition(positionMs) else { return .rejected(.positionOutOfRange) }
        return .parsed(resume ? .resume(header: header, positionMs: positionMs) : .pause(header: header, positionMs: positionMs))
    }

    private static func parseSeek(_ payload: [String: JSONValue]) -> Result {
        let header: PlaybackCommandHeader
        switch parseHeader(payload) {
        case .bad(let rejected): return rejected
        case .ok(let parsed): header = parsed
        }
        guard let target = phase5LongField(payload, fieldTargetPositionMs) else {
            return missingOrWrongType(payload, fieldTargetPositionMs)
        }
        guard isValidPosition(target) else { return .rejected(.positionOutOfRange) }
        return .parsed(.seek(header: header, targetPositionMs: target))
    }

    private static func parseStep(_ payload: [String: JSONValue], next: Bool) -> Result {
        switch parseHeader(payload) {
        case .bad(let rejected): return rejected
        case .ok(let header): return .parsed(next ? .next(header: header) : .previous(header: header))
        }
    }

    private static func parsePositionReport(_ payload: [String: JSONValue]) -> Result {
        guard let hashRaw = phase5StringField(payload, fieldTrackHash) else { return missingOrWrongType(payload, fieldTrackHash) }
        guard let trackHash = ContentHash.parse(hashRaw) else { return .rejected(.malformedContentHash) }
        guard let positionMs = phase5LongField(payload, fieldPositionMs) else { return missingOrWrongType(payload, fieldPositionMs) }
        guard isValidPosition(positionMs) else { return .rejected(.positionOutOfRange) }
        guard let atSessionUs = phase5LongField(payload, fieldAtSessionUs) else { return missingOrWrongType(payload, fieldAtSessionUs) }
        guard atSessionUs >= 0, atSessionUs <= PlaybackBounds.maxWireInt else { return .rejected(.sessionTimeOutOfRange) }
        guard let playing = phase5BoolField(payload, fieldPlaying) else { return missingOrWrongType(payload, fieldPlaying) }
        guard let rate = phase5DoubleField(payload, fieldPlaybackRate) else { return missingOrWrongType(payload, fieldPlaybackRate) }
        guard rate >= PlaybackBounds.minPlaybackRate, rate <= PlaybackBounds.maxPlaybackRate else {
            return .rejected(.rateOutOfRange)
        }
        return .parsed(
            .positionReport(trackHash: trackHash, positionMs: positionMs, atSessionUs: atSessionUs, playing: playing, playbackRate: rate)
        )
    }

    private static func parsePlaybackState(_ payload: [String: JSONValue]) -> Result {
        guard let commandSeq = phase5LongField(payload, fieldCommandSeq) else { return missingOrWrongType(payload, fieldCommandSeq) }
        guard commandSeq >= 0, commandSeq <= PlaybackBounds.maxWireInt else { return .rejected(.commandSeqOutOfRange) }
        guard let queueRevision = phase5LongField(payload, fieldQueueRevision) else {
            return missingOrWrongType(payload, fieldQueueRevision)
        }
        guard queueRevision >= 0, queueRevision <= PlaybackBounds.maxWireInt else { return .rejected(.revisionOutOfRange) }
        var trackHash: ContentHash?
        switch phase5NullableStringField(payload, fieldTrackHash) {
        case .missing: return missingOrWrongType(payload, fieldTrackHash)
        case .explicitNull: trackHash = nil
        case .present(let raw):
            guard let parsed = ContentHash.parse(raw) else { return .rejected(.malformedContentHash) }
            trackHash = parsed
        }
        var queueItemId: String?
        switch phase5NullableStringField(payload, fieldQueueItemId) {
        case .missing: return missingOrWrongType(payload, fieldQueueItemId)
        case .explicitNull: queueItemId = nil
        case .present(let raw):
            guard phase5IsUlid(raw) else { return .rejected(.malformedQueueItemId) }
            queueItemId = raw
        }
        guard let positionMs = phase5LongField(payload, fieldPositionMs) else { return missingOrWrongType(payload, fieldPositionMs) }
        guard isValidPosition(positionMs) else { return .rejected(.positionOutOfRange) }
        guard let playing = phase5BoolField(payload, fieldPlaying) else { return missingOrWrongType(payload, fieldPlaying) }
        guard let atSessionUs = phase5LongField(payload, fieldAtSessionUs) else { return missingOrWrongType(payload, fieldAtSessionUs) }
        guard atSessionUs >= 0, atSessionUs <= PlaybackBounds.maxWireInt else { return .rejected(.sessionTimeOutOfRange) }
        return .parsed(
            .playbackState(
                commandSeq: commandSeq,
                queueRevision: queueRevision,
                trackHash: trackHash,
                queueItemId: queueItemId,
                positionMs: positionMs,
                playing: playing,
                atSessionUs: atSessionUs
            )
        )
    }

    private static func isValidPosition(_ positionMs: Int64) -> Bool {
        positionMs >= 0 && positionMs <= PlaybackBounds.maxPositionMs
    }

    private static func missingOrWrongType(_ payload: [String: JSONValue], _ key: String) -> Result {
        payload[key] != nil ? .rejected(.wrongFieldType) : .rejected(.missingField)
    }
}

File: android/core/src/main/kotlin/com/ridelink/core/protocol/PlaybackCodec.kt
package com.ridelink.core.protocol

import com.ridelink.core.model.ContentHash
import com.ridelink.core.playback.PlaybackBounds
import com.ridelink.core.playback.PlaybackCommandHeader
import com.ridelink.core.playback.PlaybackMessage
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonObjectBuilder
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/** PROTOCOL §3 Playback group (Phase 5). */
object PlaybackMessageTypes {
    const val PLAY = "PLAY"
    const val PAUSE = "PAUSE"
    const val RESUME = "RESUME"
    const val SEEK = "SEEK"
    const val NEXT = "NEXT"
    const val PREVIOUS = "PREVIOUS"
    const val POSITION_REPORT = "POSITION_REPORT"
    const val PLAYBACK_STATE = "PLAYBACK_STATE"

    val ALL = setOf(PLAY, PAUSE, RESUME, SEEK, NEXT, PREVIOUS, POSITION_REPORT, PLAYBACK_STATE)

    /** The subset that carries a [PlaybackCommandHeader] and is therefore ordered by `command_seq`. */
    val COMMANDS = setOf(PLAY, PAUSE, RESUME, SEEK, NEXT, PREVIOUS)
}

/** Why a playback payload was refused. Recorded in diagnostics; never sent to the peer verbatim. */
enum class PlaybackMessageRejection {
    UNKNOWN_TYPE,
    MISSING_FIELD,
    WRONG_FIELD_TYPE,
    MALFORMED_CONTENT_HASH,
    MALFORMED_QUEUE_ITEM_ID,
    MALFORMED_PEER_ID,
    COMMAND_SEQ_OUT_OF_RANGE,
    REVISION_OUT_OF_RANGE,
    SESSION_TIME_OUT_OF_RANGE,
    POSITION_OUT_OF_RANGE,
    RATE_OUT_OF_RANGE,
}

/**
 * Parses, bounds-checks and encodes PROTOCOL §5's playback messages. Total and non-throwing,
 * mirroring [TransferCodec]/[AudioStateCodec]/[VoiceSignalCodec] — a malformed frame is dropped and
 * the control connection survives, because the framing was intact and only this message's shape was
 * wrong.
 *
 * Mirrored by `RideLinkCore.PlaybackCodec`; both run `protocol/vectors/playback-messages/`.
 */
@Suppress("TooManyFunctions")
object PlaybackCodec {
    sealed class Result {
        data class Parsed(
            val message: PlaybackMessage,
        ) : Result()

        data class Rejected(
            val reason: PlaybackMessageRejection,
        ) : Result()
    }

    const val FIELD_COMMAND_SEQ = "command_seq"
    const val FIELD_EFFECTIVE_AT_SESSION_US = "effective_at_session_us"
    const val FIELD_ISSUED_BY = "issued_by"
    const val FIELD_QUEUE_REVISION = "queue_revision"
    const val FIELD_TRACK_HASH = "track_hash"
    const val FIELD_POSITION_MS = "position_ms"
    const val FIELD_TARGET_POSITION_MS = "target_position_ms"
    const val FIELD_QUEUE_ITEM_ID = "queue_item_id"
    const val FIELD_AT_SESSION_US = "at_session_us"
    const val FIELD_PLAYING = "playing"
    const val FIELD_PLAYBACK_RATE = "playback_rate"

    fun parse(
        type: String,
        payload: JsonObject,
    ): Result =
        when (type) {
            PlaybackMessageTypes.PLAY -> parsePlay(payload)
            PlaybackMessageTypes.PAUSE -> parsePauseOrResume(payload, resume = false)
            PlaybackMessageTypes.RESUME -> parsePauseOrResume(payload, resume = true)
            PlaybackMessageTypes.SEEK -> parseSeek(payload)
            PlaybackMessageTypes.NEXT -> parseStep(payload, next = true)
            PlaybackMessageTypes.PREVIOUS -> parseStep(payload, next = false)
            PlaybackMessageTypes.POSITION_REPORT -> parsePositionReport(payload)
            PlaybackMessageTypes.PLAYBACK_STATE -> parsePlaybackState(payload)
            else -> Result.Rejected(PlaybackMessageRejection.UNKNOWN_TYPE)
        }

    fun wireType(message: PlaybackMessage): String =
        when (message) {
            is PlaybackMessage.Play -> PlaybackMessageTypes.PLAY
            is PlaybackMessage.Pause -> PlaybackMessageTypes.PAUSE
            is PlaybackMessage.Resume -> PlaybackMessageTypes.RESUME
            is PlaybackMessage.Seek -> PlaybackMessageTypes.SEEK
            is PlaybackMessage.Next -> PlaybackMessageTypes.NEXT
            is PlaybackMessage.Previous -> PlaybackMessageTypes.PREVIOUS
            is PlaybackMessage.PositionReport -> PlaybackMessageTypes.POSITION_REPORT
            is PlaybackMessage.PlaybackStateSnapshot -> PlaybackMessageTypes.PLAYBACK_STATE
        }

    /** The outbound side of [parse] — the shape lives here, once, shared by both directions. */
    fun encode(message: PlaybackMessage): JsonObject =
        when (message) {
            is PlaybackMessage.Play ->
                buildJsonObject {
                    putHeader(message.header)
                    put(FIELD_TRACK_HASH, message.trackHash.value)
                    put(FIELD_POSITION_MS, message.positionMs)
                    put(FIELD_QUEUE_ITEM_ID, message.queueItemId)
                }
            is PlaybackMessage.Pause ->
                buildJsonObject {
                    putHeader(message.header)
                    put(FIELD_POSITION_MS, message.positionMs)
                }
            is PlaybackMessage.Resume ->
                buildJsonObject {
                    putHeader(message.header)
                    put(FIELD_POSITION_MS, message.positionMs)
                }
            is PlaybackMessage.Seek ->
                buildJsonObject {
                    putHeader(message.header)
                    put(FIELD_TARGET_POSITION_MS, message.targetPositionMs)
                }
            is PlaybackMessage.Next -> buildJsonObject { putHeader(message.header) }
            is PlaybackMessage.Previous -> buildJsonObject { putHeader(message.header) }
            is PlaybackMessage.PositionReport ->
                buildJsonObject {
                    put(FIELD_TRACK_HASH, message.trackHash.value)
                    put(FIELD_POSITION_MS, message.positionMs)
                    put(FIELD_AT_SESSION_US, message.atSessionUs)
                    put(FIELD_PLAYING, message.playing)
                    put(FIELD_PLAYBACK_RATE, message.playbackRate)
                }
            is PlaybackMessage.PlaybackStateSnapshot ->
                buildJsonObject {
                    put(FIELD_COMMAND_SEQ, message.commandSeq)
                    put(FIELD_QUEUE_REVISION, message.queueRevision)
                    put(FIELD_TRACK_HASH, message.trackHash?.value)
                    put(FIELD_QUEUE_ITEM_ID, message.queueItemId)
                    put(FIELD_POSITION_MS, message.positionMs)
                    put(FIELD_PLAYING, message.playing)
                    put(FIELD_AT_SESSION_US, message.atSessionUs)
                }
        }

    private fun JsonObjectBuilder.putHeader(header: PlaybackCommandHeader) {
        put(FIELD_COMMAND_SEQ, header.commandSeq)
        put(FIELD_EFFECTIVE_AT_SESSION_US, header.effectiveAtSessionUs)
        put(FIELD_ISSUED_BY, header.issuedBy.value)
        put(FIELD_QUEUE_REVISION, header.queueRevision)
    }

    // --- headers -------------------------------------------------------------------------------

    private sealed class HeaderResult {
        data class Ok(
            val header: PlaybackCommandHeader,
        ) : HeaderResult()

        data class Bad(
            val rejected: Result.Rejected,
        ) : HeaderResult()
    }

    @Suppress("ReturnCount")
    private fun parseHeader(payload: JsonObject): HeaderResult {
        val commandSeq =
            phase5LongField(payload, FIELD_COMMAND_SEQ)
                ?: return HeaderResult.Bad(missingOrWrongType(payload, FIELD_COMMAND_SEQ))
        if (commandSeq < 0 || commandSeq > PlaybackBounds.MAX_WIRE_INT) {
            return HeaderResult.Bad(Result.Rejected(PlaybackMessageRejection.COMMAND_SEQ_OUT_OF_RANGE))
        }
        val effectiveAt =
            phase5LongField(payload, FIELD_EFFECTIVE_AT_SESSION_US)
                ?: return HeaderResult.Bad(missingOrWrongType(payload, FIELD_EFFECTIVE_AT_SESSION_US))
        if (effectiveAt < 0 || effectiveAt > PlaybackBounds.MAX_WIRE_INT) {
            return HeaderResult.Bad(Result.Rejected(PlaybackMessageRejection.SESSION_TIME_OUT_OF_RANGE))
        }
        val issuedByRaw =
            phase5StringField(payload, FIELD_ISSUED_BY)
                ?: return HeaderResult.Bad(missingOrWrongType(payload, FIELD_ISSUED_BY))
        val issuedBy =
            phase5ParsePeerId(issuedByRaw) ?: return HeaderResult.Bad(Result.Rejected(PlaybackMessageRejection.MALFORMED_PEER_ID))
        val queueRevision =
            phase5LongField(payload, FIELD_QUEUE_REVISION) ?: return HeaderResult.Bad(missingOrWrongType(payload, FIELD_QUEUE_REVISION))
        if (queueRevision < 0 || queueRevision > PlaybackBounds.MAX_WIRE_INT) {
            return HeaderResult.Bad(Result.Rejected(PlaybackMessageRejection.REVISION_OUT_OF_RANGE))
        }
        return HeaderResult.Ok(PlaybackCommandHeader(commandSeq, effectiveAt, issuedBy, queueRevision))
    }

    // --- per-type parsers ----------------------------------------------------------------------

    @Suppress("ReturnCount")
    private fun parsePlay(payload: JsonObject): Result {
        val header =
            when (val h = parseHeader(payload)) {
                is HeaderResult.Bad -> return h.rejected
                is HeaderResult.Ok -> h.header
            }
        val hashRaw = phase5StringField(payload, FIELD_TRACK_HASH) ?: return missingOrWrongType(payload, FIELD_TRACK_HASH)
        val trackHash = ContentHash.parse(hashRaw) ?: return Result.Rejected(PlaybackMessageRejection.MALFORMED_CONTENT_HASH)
        val positionMs = phase5LongField(payload, FIELD_POSITION_MS) ?: return missingOrWrongType(payload, FIELD_POSITION_MS)
        if (!isValidPosition(positionMs)) return Result.Rejected(PlaybackMessageRejection.POSITION_OUT_OF_RANGE)
        val queueItemId = phase5StringField(payload, FIELD_QUEUE_ITEM_ID) ?: return missingOrWrongType(payload, FIELD_QUEUE_ITEM_ID)
        if (!phase5IsUlid(queueItemId)) return Result.Rejected(PlaybackMessageRejection.MALFORMED_QUEUE_ITEM_ID)
        return Result.Parsed(PlaybackMessage.Play(header, trackHash, positionMs, queueItemId))
    }

    @Suppress("ReturnCount")
    private fun parsePauseOrResume(
        payload: JsonObject,
        resume: Boolean,
    ): Result {
        val header =
            when (val h = parseHeader(payload)) {
                is HeaderResult.Bad -> return h.rejected
                is HeaderResult.Ok -> h.header
            }
        val positionMs = phase5LongField(payload, FIELD_POSITION_MS) ?: return missingOrWrongType(payload, FIELD_POSITION_MS)
        if (!isValidPosition(positionMs)) return Result.Rejected(PlaybackMessageRejection.POSITION_OUT_OF_RANGE)
        return Result.Parsed(
            if (resume) PlaybackMessage.Resume(header, positionMs) else PlaybackMessage.Pause(header, positionMs),
        )
    }

    @Suppress("ReturnCount")
    private fun parseSeek(payload: JsonObject): Result {
        val header =
            when (val h = parseHeader(payload)) {
                is HeaderResult.Bad -> return h.rejected
                is HeaderResult.Ok -> h.header
            }
        val target = phase5LongField(payload, FIELD_TARGET_POSITION_MS) ?: return missingOrWrongType(payload, FIELD_TARGET_POSITION_MS)
        if (!isValidPosition(target)) return Result.Rejected(PlaybackMessageRejection.POSITION_OUT_OF_RANGE)
        return Result.Parsed(PlaybackMessage.Seek(header, target))
    }

    private fun parseStep(
        payload: JsonObject,
        next: Boolean,
    ): Result =
        when (val h = parseHeader(payload)) {
            is HeaderResult.Bad -> h.rejected
            is HeaderResult.Ok ->
                Result.Parsed(if (next) PlaybackMessage.Next(h.header) else PlaybackMessage.Previous(h.header))
        }

    @Suppress("ReturnCount")
    private fun parsePositionReport(payload: JsonObject): Result {
        val hashRaw = phase5StringField(payload, FIELD_TRACK_HASH) ?: return missingOrWrongType(payload, FIELD_TRACK_HASH)
        val trackHash = ContentHash.parse(hashRaw) ?: return Result.Rejected(PlaybackMessageRejection.MALFORMED_CONTENT_HASH)
        val positionMs = phase5LongField(payload, FIELD_POSITION_MS) ?: return missingOrWrongType(payload, FIELD_POSITION_MS)
        if (!isValidPosition(positionMs)) return Result.Rejected(PlaybackMessageRejection.POSITION_OUT_OF_RANGE)
        val atSessionUs = phase5LongField(payload, FIELD_AT_SESSION_US) ?: return missingOrWrongType(payload, FIELD_AT_SESSION_US)
        if (atSessionUs < 0 || atSessionUs > PlaybackBounds.MAX_WIRE_INT) {
            return Result.Rejected(PlaybackMessageRejection.SESSION_TIME_OUT_OF_RANGE)
        }
        val playing = phase5BooleanField(payload, FIELD_PLAYING) ?: return missingOrWrongType(payload, FIELD_PLAYING)
        val rate = phase5DoubleField(payload, FIELD_PLAYBACK_RATE) ?: return missingOrWrongType(payload, FIELD_PLAYBACK_RATE)
        if (rate < PlaybackBounds.MIN_PLAYBACK_RATE || rate > PlaybackBounds.MAX_PLAYBACK_RATE) {
            return Result.Rejected(PlaybackMessageRejection.RATE_OUT_OF_RANGE)
        }
        return Result.Parsed(PlaybackMessage.PositionReport(trackHash, positionMs, atSessionUs, playing, rate))
    }

    @Suppress("ReturnCount", "CyclomaticComplexMethod")
    private fun parsePlaybackState(payload: JsonObject): Result {
        val commandSeq = phase5LongField(payload, FIELD_COMMAND_SEQ) ?: return missingOrWrongType(payload, FIELD_COMMAND_SEQ)
        if (commandSeq < 0 || commandSeq > PlaybackBounds.MAX_WIRE_INT) {
            return Result.Rejected(PlaybackMessageRejection.COMMAND_SEQ_OUT_OF_RANGE)
        }
        val queueRevision = phase5LongField(payload, FIELD_QUEUE_REVISION) ?: return missingOrWrongType(payload, FIELD_QUEUE_REVISION)
        if (queueRevision < 0 || queueRevision > PlaybackBounds.MAX_WIRE_INT) {
            return Result.Rejected(PlaybackMessageRejection.REVISION_OUT_OF_RANGE)
        }
        val trackHash =
            when (val raw = phase5NullableStringField(payload, FIELD_TRACK_HASH)) {
                Phase5NullableString.Missing -> return missingOrWrongType(payload, FIELD_TRACK_HASH)
                Phase5NullableString.ExplicitNull -> null
                is Phase5NullableString.Present ->
                    ContentHash.parse(raw.value) ?: return Result.Rejected(PlaybackMessageRejection.MALFORMED_CONTENT_HASH)
            }
        val queueItemId =
            when (val raw = phase5NullableStringField(payload, FIELD_QUEUE_ITEM_ID)) {
                Phase5NullableString.Missing -> return missingOrWrongType(payload, FIELD_QUEUE_ITEM_ID)
                Phase5NullableString.ExplicitNull -> null
                is Phase5NullableString.Present ->
                    if (phase5IsUlid(raw.value)) raw.value else return Result.Rejected(PlaybackMessageRejection.MALFORMED_QUEUE_ITEM_ID)
            }
        val positionMs = phase5LongField(payload, FIELD_POSITION_MS) ?: return missingOrWrongType(payload, FIELD_POSITION_MS)
        if (!isValidPosition(positionMs)) return Result.Rejected(PlaybackMessageRejection.POSITION_OUT_OF_RANGE)
        val playing = phase5BooleanField(payload, FIELD_PLAYING) ?: return missingOrWrongType(payload, FIELD_PLAYING)
        val atSessionUs = phase5LongField(payload, FIELD_AT_SESSION_US) ?: return missingOrWrongType(payload, FIELD_AT_SESSION_US)
        if (atSessionUs < 0 || atSessionUs > PlaybackBounds.MAX_WIRE_INT) {
            return Result.Rejected(PlaybackMessageRejection.SESSION_TIME_OUT_OF_RANGE)
        }
        return Result.Parsed(
            PlaybackMessage.PlaybackStateSnapshot(commandSeq, queueRevision, trackHash, queueItemId, positionMs, playing, atSessionUs),
        )
    }

    private fun isValidPosition(positionMs: Long): Boolean = positionMs >= 0 && positionMs <= PlaybackBounds.MAX_POSITION_MS

    private fun missingOrWrongType(
        payload: JsonObject,
        key: String,
    ): Result.Rejected =
        if (payload.containsKey(key)) {
            Result.Rejected(PlaybackMessageRejection.WRONG_FIELD_TYPE)
        } else {
            Result.Rejected(PlaybackMessageRejection.MISSING_FIELD)
        }
}
